DQMOffline/Trigger/plugins/BPHMonitor.cc

```diff
@@ -146,55 +314,334 @@ void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
-    else if (m.isStandAloneMuon()) track = & * m.outerTrack();
-    if (track) {
-      double d0 = track->dxy(beamSpot->position());
-      double z0 = track->dz(beamSpot->position());
-      mud0_.denominator ->Fill(d0);
-      muz0_.denominator ->Fill(z0);
-    }
-  } 
-  // applying selection for numerator
+//  edm::Handle<reco::BeamSpot> const& beamSpot;
+  // Filter out events if Trigger Filtering is requested
+  if (tnp_>0) {
   if (num_genTriggerEventFlag_->on() && ! num_genTriggerEventFlag_->accept( iEvent, iSetup) ) return;
-  for (auto const & m : *muoHandle) {
-    if(!muoSelection_(m))continue;
-    muPhi_.numerator->Fill(m.phi());
-    muEta_.numerator->Fill(m.eta());
-    muPt_.numerator ->Fill(m.pt());
-    const reco::Track * track = 0;
-    if (m.isTrackerMuon()) track = & * m.innerTrack();
-    else if (m.isStandAloneMuon()) track = & * m.outerTrack();
-    if (track) {
-      double d0 = track->dxy(beamSpot->position());
-      double z0 = track->dz(beamSpot->position());
-      mud0_.numerator ->Fill(d0);
-      muz0_.numerator ->Fill(z0);
+  std::vector<reco::Muon> tagMuons;
+  std::vector<reco::Muon> probeMuons;
+  for ( auto const & m : *muoHandle ) {
+  if ( muoSelection_tag( m ) ) tagMuons.push_back(m);
+  }
+  for (int i = 0; i<int(tagMuons.size());i++){
+    for ( auto const & m : *muoHandle ) { 
+      if ((tagMuons[i].pt() - m.pt())<=0.01)continue;//not the same  
+      if ((tagMuons[i].p4()+m.p4()).M() >2.596&& (tagMuons[i].p4()+m.p4()).M() <3.596){//near to J/psi mass
+      muPhi_.denominator->Fill(m.phi());
+      muEta_.denominator->Fill(m.eta());
+      muPt_.denominator ->Fill(m.pt());
+      if (muoSelection_probe( m )){
+        muPhi_.numerator->Fill(m.phi());
+        muEta_.numerator->Fill(m.eta());
+        muPt_.numerator ->Fill(m.pt());
+
+        }
+
+      }      
+
     }
+      
+  }
+    
+
+  }  
+  else{
+    if (trOrMu_){
+      edm::Handle<reco::TrackCollection> trHandle;
+      iEvent.getByToken( trToken_, trHandle );
+  for (auto const & t : *trHandle) {
+    if(!trSelection_ref(t))continue;
+    muPhi_.denominator->Fill(t.phi());
+    muEta_.denominator->Fill(t.eta());
+    muPt_.denominator ->Fill(t.pt());
+  }
+  if (num_genTriggerEventFlag_->on() && ! num_genTriggerEventFlag_->accept( iEvent, iSetup) ) return;
+  for (auto const & t : *trHandle) {
+    if(!trSelection_(t))continue;
+    muPhi_.numerator->Fill(t.phi());
+    muEta_.numerator->Fill(t.eta());
+    muPt_.numerator ->Fill(t.pt());
   }
-	
-  // filling histograms (num_genTriggerEventFlag_)  
 
+//  
+
+  
+    }
+
+  else{
+  if (den_genTriggerEventFlag_->on() && ! den_genTriggerEventFlag_->accept( iEvent, iSetup) ) return;
+  for (auto const & m : *muoHandle ) {
+  for (auto const & m1 : *muoHandle ) {
+        
+        if (m1.pt() == m.pt())continue;
+      if(!muoSelection_ref(m))continue;   
+      if(!muoSelection_ref(m1))continue;   
+      switch(nofset_){
+      case 1: tnp_=1;
```

Comment:
These cases are for different sets of variables to monitor the efficiency. 
They are listed in this table, nofset variable(monitoring code) column
https://docs.google.com/spreadsheets/d/1VkVihttps://docs.google.com/spreadsheets/d/1VkVioMr63WIEV-KWORUQpjmGdSUVsH3cx0VoREDO1VM/edit#gid=0oMr63WIEV-KWORUQpjmGdSUVsH3cx0VoREDO1VM/edit#gid=0